text/tls-and-alpn.rst

```diff
@@ -0,0 +1,380 @@
+parameters.
+
+
+CLI Tooling
+-----------
+
+For the ``edgedb`` CLI, the goal is to make TLS in EdgeDB easy to use. There
+are 2 major scenarios:
+
+1. Using EdgeDB with TLS in development process.
+2. Deploy EdgeDB with TLS for production environment.
+
+As we'd like to enable TLS by default, a certificate is always required before
+starting an EdgeDB instance. In brief, for development, a self-signed
+certificate is usually sufficient, so the CLI tooling should have an option to
+automatically generate that for the user transparently. For production, we'd
+also like to make the certificate/key management easier and safe, options are
+discussed later.
+
+
+Internals
+^^^^^^^^^
+
+Under the neath, we'll have new files and folders in the EdgeDB well-known
+folder::
+
+    $WELLKNOWN/
+        +--config/
+        |    +--instance_ports.json
+        |    +--tls.json              <NEW>
+        |--credentials/
+        |    +--<instance>.json       <UPDATED>
+        |--certs/                     <NEW>
+        |    +--<instance>.cert.pem   <NEW>
+        |    +--<instance>.key.pem    <NEW>
+        ...
+
+The ``config/tls.json`` contains optional TLS settings for each instance::
+
+    {
+        "<instance>": {
+            "protocol_min": "<min TLS version>",
+            "protocol_max": "<max TLS version>",
+            "ciphers": "<available ciphers>"
+        },
+        ...
+    }
+
+If an instance has no record in ``config/tls.json``, the EdgeDB server options
+``--tls-protocol-min``, ``--tls-protocol-max`` and ``--tls-ciphers`` will be
+omitted and the default value will be used.
+
+``credentials/<instance>.json`` is updated to include the passphrase for the
+private key file, if any. For example::
+
+    {
+        "port": 10732,
+        "user": "edgedb",
+        "password": "....",
+        "database": "edgedb",
+        "tls_private_key_password": "...."
+    }
+
+Question: I don't feel comfortable storing the key passphrase here in clear
+text, especially if this is in production environment and we asked the user
+for passphrase but stored in this file. Perhaps we could use this JSON value
+only for the development process, and introduce a more secure (thus less
+convenient) way to obtain the private key password, like a customizable command
+as in PostgreSQL?
+
+Under the new folder ``certs/``, we'll place all certificates and keys used by
+the EdgeDB instances. The cert files will be named after the instance names.
+For development scenario we'll likely generate those certs for the user, while
+in production we'll "import" the certs specified by the user, by copying and
+renaming the cert and key files to the ``certs/`` folder.
+
+
+Development Scenario
+^^^^^^^^^^^^^^^^^^^^
+
+From the user's perspective, ``edgedb server init`` (without parameters we are
+introducing in this RFC later) will now automatically generate a self-signed
+certificate with a random key passphrase, and set the above config properly.
```

Comment:
Perhaps we should ask the user to enter the passphrase (or supply it via a command line arg or an env var)

Reply:
For the self-signed certificate? Sure we can have the 3 ways to collect a passphrase, but I'm not quite getting the idea to have passphrase for development 🤔

Reply:
I agree, we don't need a passphrase here.